rspace/src/main/scala/coop/rchain/rspace/history/InMemoryTrieStore.scala

```diff
@@ -0,0 +1,117 @@
+    State(_dbTrie, _dbRoot, newPastRoots)
+}
+
+object State {
+  def empty[K, V]: State[K, V] = State[K, V](Map.empty, Map.empty, Map.empty)
+}
+
+class InMemoryTrieStore[K, V]
+    extends InMemoryOps[State[K, V]]
+    with ITrieStore[InMemTransaction[State[K, V]], K, V] {
+
+  override def emptyState: State[K, V] = State.empty
+
+  private[this] val refine       = Map("path" -> "inmemTrie")
+  private[this] val entriesGauge = Kamon.gauge("entries").refine(refine)
+
+  override private[rspace] def updateGauges(): Unit =
+    withTxn(createTxnRead())(_.readState { state =>
+      entriesGauge.set(state._dbTrie.size.toLong)
+    })
+
+  override private[rspace] def getRoot(txn: InMemTransaction[State[K, V]],
+                                       branch: Branch): Option[Blake2b256Hash] =
+    txn.readState(state => state._dbRoot.get(branch))
+
+  override private[rspace] def persistAndGetRoot(txn: InMemTransaction[State[K, V]],
+                                                 branch: Branch): Option[Blake2b256Hash] =
+    getRoot(txn, branch)
+      .map { currentRoot =>
+        val pastRoots = getPastRootsInBranch(txn, branch).filter(_ != currentRoot)
+        (currentRoot, currentRoot +: pastRoots)
+      }
+      .map {
+        case (currentRoot, updatedPastRoots) =>
+          txn.writeState(state =>
+            (state.chnagePastRoot(state._dbPastRoots + (branch -> updatedPastRoots)), ()))
+          currentRoot
+      }
+
+  private[this] def getPastRootsInBranch(txn: InMemTransaction[State[K, V]],
+                                         branch: Branch): Seq[Blake2b256Hash] =
+    txn.readState(state => state._dbPastRoots.getOrElse(branch, Seq.empty))
+
+  override private[rspace] def putRoot(txn: InMemTransaction[State[K, V]],
+                                       branch: Branch,
+                                       hash: Blake2b256Hash): Unit =
+    txn.writeState(state => (state.changeRoot(state._dbRoot + (branch -> hash)), ()))
+
+  override private[rspace] def getAllPastRoots(
+      txn: InMemTransaction[State[K, V]]): Seq[Blake2b256Hash] =
+    txn.readState(state =>
+      state._dbPastRoots.values.foldLeft(Seq.empty[Blake2b256Hash])((acc, value) => acc ++ value))
+
+  override private[rspace] def validateAndPutRoot(txn: InMemTransaction[State[K, V]],
+                                                  branch: Branch,
+                                                  hash: Blake2b256Hash): Unit =
+    getRoot(txn, branch)
+      .find(_ == hash)
+      .orElse {
+        getPastRootsInBranch(txn, branch)
+          .find(_ == hash)
+          .map { blake: Blake2b256Hash =>
+            putRoot(txn, branch, blake)
+            blake
+          }
+      }
+      .orElse {
+        getAllPastRoots(txn)
+          .find(_ == hash)
+          .map { blake: Blake2b256Hash =>
+            putRoot(txn, branch, blake)
+            blake
+          }
+      }
+      .getOrElse(throw new Exception(s"Unknown root."))
```

Comment:
I understand the need to keep the implementation close, but don't you think this could be just the last `else`? It's all in memory so it should be good enough.

Reply:
Sure, updated